File: Go/EX4.go
package main

import (
	"fmt"
)

// Interface Animal
type Animal interface {
	Som() string
}

// Struct Cachorro
type Cachorro struct{}

// Implementação do método Som para Cachorro
func (c Cachorro) Som() string {
	return "Au Au"
}

// Struct Gato
type Gato struct{}

// Implementação do método Som para Gato
func (g Gato) Som() string {
	return "Miau"
}

// Função principal
func main() {
	// Criando instâncias das structs
	var cachorro Animal = Cachorro{}
	var gato Animal = Gato{}

	// Exibindo os sons dos animais
	fmt.Println("Cachorro:", cachorro.Som())
	fmt.Println("Gato:", gato.Som())
}

File: Java/EX4.java
// Classe base Animal
abstract class Animal {
    // Método abstrato para ser implementado pelas subclasses
    public abstract String som();
}

// Classe derivada Cachorro
class Cachorro extends Animal {
    @Override
    public String som() {
        return "Au Au";
    }
}

// Classe derivada Gato
class Gato extends Animal {
    @Override
    public String som() {
        return "Miau";
    }
}

// Classe principal
class Main {
    public static void main(String[] args) {
        // Criando instâncias das classes
        Animal cachorro = new Cachorro();
        Animal gato = new Gato();

        // Exibindo os sons dos animais
        System.out.println("Cachorro: " + cachorro.som());
        System.out.println("Gato: " + gato.som());
    }
}